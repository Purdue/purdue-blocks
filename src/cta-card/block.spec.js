// cta card test
jest.setTimeout(30000)
import {
    createNewPost,
    enablePageDialogAccept,
    getEditedPostContent,
    insertBlock,
} from '@wordpress/e2e-test-utils';

import {
    clickElementByText,
    blockStartup,
    clickRadio,
    clickCheckbox,
    selectOption
} from '../test-helpers'

const block = {blockTitle: 'CTA Card', blockName: 'purdue-blocks/cta-card'}

describe( 'ðŸ”¬ CTA Card Block', () => {
    beforeAll( async () => {
        await enablePageDialogAccept();
    } );
    beforeEach( async () => {
        await createNewPost();
    } );

    test( 'ðŸ”¬ Block should be available.', async () => {
        await insertBlock( block.blockTitle )

        // tests that the block is properly inserted and matches the existing snapshot
        expect(await page.$(`[data-type="${block.blockName}"]`)).not.toBeNull()
        expect( await getEditedPostContent()).toMatchSnapshot()
    } )

    describe( 'ðŸ”¬ Block Editor Fields', () => {
        test( 'ðŸ”Ž Button should open Media Library.', async () => {
            await blockStartup(block)

            // open media library
            await clickElementByText('button', 'Select an image')

            // tests that media library does open
            expect(await page.$(`div.media-frame-title`)).not.toBeNull()
        })
        describe('ðŸ”¬ Card Content Richtext fields', () => {
            test( 'ðŸ”Ž Title Richtext field.', async () => {
                await blockStartup(block)
    
                const typeString = "Test Title"
                
                // focus the text box then type into it with the virtual keyboard
                await page.focus( '.block-editor-rich-text__editable.title' );
                await page.keyboard.type(typeString, {delay: 10})
    
                const editedContent = await getEditedPostContent()
    
                expect( editedContent).toMatchSnapshot()
            })
    
            test( 'ðŸ”Ž Text Richtext field.', async () => {
                await blockStartup(block)
    
                const typeString = "Test text here."
                
                // focus the textarea then type into it with the virtual keyboard
                await page.focus( '.block-editor-rich-text__editable.content' );
                await page.keyboard.type(typeString, {delay: 10})
    
                const editedContent = await getEditedPostContent()
    
                expect( editedContent).toMatchSnapshot()
                expect( editedContent.includes(`<p class="content">${typeString}</p>`)).toBe(true)
            })
        })
    })
    describe( 'ðŸ”¬ Side Panel Settings', () => {
        describe('ðŸ”¬ General Card Settings', () => {
            test( 'ðŸ”Ž Card Size Options Radio Buttons.', async () => {
                await blockStartup(block)
    
                // select the radio button with the virtual keyboard
                await clickRadio('Card type', 'large')
    
                const editedContent = await getEditedPostContent()
    
                expect( editedContent).toMatchSnapshot()
                expect( editedContent.includes(`"cardType":"large"`)).toBe(true)
            })

            test( 'Heading level dropdown selector', async () => {
                await blockStartup(block)
    
                await selectOption('Heading level of the title', 'h2')
                let editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h2"`)).toBe(true)
    
                await selectOption('Heading level of the title', 'h3')
                editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h3"`)).toBe(true)
    
                await selectOption('Heading level of the title', 'h4')
                editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h4"`)).toBe(true)
    
                await selectOption('Heading level of the title', 'h5')
                editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h5"`)).toBe(true)
    
                await selectOption('Heading level of the title', 'h6')
                editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h6"`)).toBe(true)
    
                await selectOption('Heading level of the title', 'p')
                editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"titleLevel":"h2"`)).toBe(false)
                expect( editedContent.includes(`"titleLevel":"h3"`)).toBe(false)
                expect( editedContent.includes(`"titleLevel":"h4"`)).toBe(false)
                expect( editedContent.includes(`"titleLevel":"h5"`)).toBe(false)
                expect( editedContent.includes(`"titleLevel":"h6"`)).toBe(false)
    
            })
        })
        describe('ðŸ”¬ Card Image Settings', () => {
            test( 'ðŸ”Ž Image Alignment Options Radio Buttons', async () => {
                await blockStartup(block)
            
                // select the radio button with the virtual keyboard
                await clickRadio('Align image', 'right')
    
                const editedContent = await getEditedPostContent()
    
                expect( editedContent).toMatchSnapshot()
                expect( editedContent.includes(`"imgLocation":"right"`)).toBe(true)
            })
            test( 'ðŸ”Ž Image alt text field', async () => {
                await blockStartup(block)
    
                const typeString = "Image alt text."
                
                // focus the text box then type into it with the virtual keyboard
                await clickElementByText('label', 'Image Alt Text')
                await page.keyboard.type(typeString, {delay: 10})
    
                const editedContent = await getEditedPostContent()
    
                expect( editedContent).toMatchSnapshot()
                expect( editedContent.includes(`"altText":"${typeString}"`)).toBe(true)
            })
        })
        describe( 'ðŸ”¬ Link can be added to the card', () => {
            test( 'ðŸ”Ž "Add a link to this card?" checkbox reveals link inputs', async () => {
                await blockStartup(block)

                await clickCheckbox('Add a link to this card?')
                      
                let editedContent = await getEditedPostContent()
                
                expect( editedContent.includes('{"hasLink":true}')).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
 
                await clickCheckbox('Add a link to this card?')
                
                editedContent = await getEditedPostContent()
                
                expect( editedContent.includes('{"hasLink":true}')).toBe(false)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
            test('ðŸ”Ž Link text field updates output correctly', async () => {
                await blockStartup(block)

                const typeString = "CTA text"
                await clickCheckbox('Add a link to this card?')
                await clickElementByText('label', 'Call to action text')
                await page.keyboard.type(typeString, {delay: 10})

                const editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"linkText":"${typeString}"`)).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
            test('ðŸ”Ž Link url field updates output correctly', async () => {
                await blockStartup(block)
                await clickCheckbox('Add a link to this card?')
                const typeString = "https://www.purdue.edu"

                await clickElementByText('label', 'Link address')
                await page.keyboard.type(typeString, {delay: 10})

                const editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"link":"${typeString}"`)).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
            test('ðŸ”Ž Link can be opened in new tab', async () => {
                await blockStartup(block)

                const typeString = "https://www.purdue.edu"
                await clickCheckbox('Add a link to this card?')
                await clickElementByText('label', 'Link address')
                await page.keyboard.type(typeString, {delay: 10})

                await clickCheckbox('Open link in new tab?')

                const editedContent = await getEditedPostContent()
                expect( editedContent.includes(`"external":true`)).toBe(true)
                expect( await getEditedPostContent() ).toMatchSnapshot()
            })
        })
    })
} );
