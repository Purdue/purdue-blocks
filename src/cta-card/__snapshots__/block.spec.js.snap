// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`🔬 CTA Card Block 🔬 Block Editor Fields 🔬 Card Content Richtext fields 🔎 Text Richtext field. 1`] = `
"<!-- wp:purdue-blocks/cta-card -->
<div class=\\" cta-card-horizonal cta-card-small cta-card-left\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-two-fifths-desktop is-two-fifths-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"column is-three-fifths-desktop is-three-fifths-tablet is-full-mobile\\"><p class=\\"content\\">Test text here.</p></div></div></div>
<!-- /wp:purdue-blocks/cta-card -->"
`;

exports[`🔬 CTA Card Block 🔬 Block Editor Fields 🔬 Card Content Richtext fields 🔎 Title Richtext field. 1`] = `
"<!-- wp:purdue-blocks/cta-card -->
<div class=\\" cta-card-horizonal cta-card-small cta-card-left\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-two-fifths-desktop is-two-fifths-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"column is-three-fifths-desktop is-three-fifths-tablet is-full-mobile\\"><p class=\\"title\\">Test Title</p></div></div></div>
<!-- /wp:purdue-blocks/cta-card -->"
`;

exports[`🔬 CTA Card Block 🔬 Block should be available. 1`] = `
"<!-- wp:purdue-blocks/cta-card -->
<div class=\\" cta-card-horizonal cta-card-small cta-card-left\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-two-fifths-desktop is-two-fifths-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"column is-three-fifths-desktop is-three-fifths-tablet is-full-mobile\\"></div></div></div>
<!-- /wp:purdue-blocks/cta-card -->"
`;

exports[`🔬 CTA Card Block 🔬 Side Panel Settings 🔬 Card Image Settings 🔎 Image Alignment Options Radio Buttons 1`] = `
"<!-- wp:purdue-blocks/cta-card {\\"imgLocation\\":\\"right\\"} -->
<div class=\\" cta-card-horizonal cta-card-small cta-card-right\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-two-fifths-desktop is-two-fifths-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"column is-three-fifths-desktop is-three-fifths-tablet is-full-mobile\\"></div></div></div>
<!-- /wp:purdue-blocks/cta-card -->"
`;

exports[`🔬 CTA Card Block 🔬 Side Panel Settings 🔬 Card Image Settings 🔎 Image alt text field 1`] = `
"<!-- wp:purdue-blocks/cta-card {\\"altText\\":\\"Image alt text.\\"} -->
<div class=\\" cta-card-horizonal cta-card-small cta-card-left\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-two-fifths-desktop is-two-fifths-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"Image alt text.\\"/></figure></div><div class=\\"column is-three-fifths-desktop is-three-fifths-tablet is-full-mobile\\"></div></div></div>
<!-- /wp:purdue-blocks/cta-card -->"
`;

exports[`🔬 CTA Card Block 🔬 Side Panel Settings 🔬 General Card Settings 🔎 Card Size Options Radio Buttons. 1`] = `
"<!-- wp:purdue-blocks/cta-card {\\"cardType\\":\\"large\\"} -->
<div class=\\" cta-card-horizonal cta-card-large cta-card-left\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-one-third-desktop is-one-third-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"column is-two-thirds-desktop is-two-thirds-tablet is-full-mobile\\"><!-- wp:paragraph {\\"placeholder\\":\\"Add content\\"} -->
<p></p>
<!-- /wp:paragraph --></div></div></div>
<!-- /wp:purdue-blocks/cta-card -->"
`;

exports[`🔬 CTA Card Block 🔬 Side Panel Settings 🔬 Link can be added to the card 🔎 "Add a link to this card?" checkbox reveals link inputs 1`] = `
"<!-- wp:purdue-blocks/cta-card {\\"hasLink\\":true} -->
<a href=\\"\\" target=\\"_self\\" class=\\"cta-card-horizonal cta-card-small cta-card-left\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-two-fifths-desktop is-two-fifths-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"column is-three-fifths-desktop is-three-fifths-tablet is-full-mobile\\"></div></div></a>
<!-- /wp:purdue-blocks/cta-card -->"
`;

exports[`🔬 CTA Card Block 🔬 Side Panel Settings 🔬 Link can be added to the card 🔎 "Add a link to this card?" checkbox reveals link inputs 2`] = `
"<!-- wp:purdue-blocks/cta-card -->
<div class=\\" cta-card-horizonal cta-card-small cta-card-left\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-two-fifths-desktop is-two-fifths-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"column is-three-fifths-desktop is-three-fifths-tablet is-full-mobile\\"></div></div></div>
<!-- /wp:purdue-blocks/cta-card -->"
`;

exports[`🔬 CTA Card Block 🔬 Side Panel Settings 🔬 Link can be added to the card 🔎 Link can be opened in new tab 1`] = `
"<!-- wp:purdue-blocks/cta-card {\\"hasLink\\":true,\\"link\\":\\"https://www.purdue.edu\\",\\"external\\":true} -->
<a href=\\"https://www.purdue.edu\\" target=\\"_blank\\" class=\\"cta-card-horizonal cta-card-small cta-card-left\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-two-fifths-desktop is-two-fifths-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"column is-three-fifths-desktop is-three-fifths-tablet is-full-mobile\\"></div></div></a>
<!-- /wp:purdue-blocks/cta-card -->"
`;

exports[`🔬 CTA Card Block 🔬 Side Panel Settings 🔬 Link can be added to the card 🔎 Link text field updates output correctly 1`] = `
"<!-- wp:purdue-blocks/cta-card {\\"hasLink\\":true,\\"linkText\\":\\"CTA text\\"} -->
<a href=\\"\\" target=\\"_self\\" class=\\"cta-card-horizonal cta-card-small cta-card-left\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-two-fifths-desktop is-two-fifths-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"column is-three-fifths-desktop is-three-fifths-tablet is-full-mobile\\"><div class=\\"read-more-button\\"><span>CTA text</span></div></div></div></a>
<!-- /wp:purdue-blocks/cta-card -->"
`;

exports[`🔬 CTA Card Block 🔬 Side Panel Settings 🔬 Link can be added to the card 🔎 Link url field updates output correctly 1`] = `
"<!-- wp:purdue-blocks/cta-card {\\"hasLink\\":true,\\"link\\":\\"https://www.purdue.edu\\"} -->
<a href=\\"https://www.purdue.edu\\" target=\\"_self\\" class=\\"cta-card-horizonal cta-card-small cta-card-left\\" rel=\\"noopener noreferrer\\"><div class=\\"columns is-multiline\\"><div class=\\"column is-two-fifths-desktop is-two-fifths-tablet is-full-mobile\\"><figure class=\\"image is-3by2\\"><img src=\\"\\" alt=\\"\\"/></figure></div><div class=\\"column is-three-fifths-desktop is-three-fifths-tablet is-full-mobile\\"></div></div></a>
<!-- /wp:purdue-blocks/cta-card -->"
`;
